src/device/rcp/rsp/rsp_core.c

```diff
@@ -34,60 +34,93 @@
 #include "main/profile.h"
 #endif
 #include "plugin/plugin.h"
+#include "api/callbacks.h"
 
-static void dma_sp_write(struct rsp_core* sp)
+static void do_sp_dma(struct rsp_core* sp, struct sp_dma* dma)
 {
     unsigned int i,j;
 
-    unsigned int l = sp->regs[SP_RD_LEN_REG];
+    unsigned int l = dma->length;
 
     unsigned int length = ((l & 0xfff) | 7) + 1;
     unsigned int count = ((l >> 12) & 0xff) + 1;
     unsigned int skip = ((l >> 20) & 0xfff);
 
-    unsigned int memaddr = sp->regs[SP_MEM_ADDR_REG] & 0xfff;
-    unsigned int dramaddr = sp->regs[SP_DRAM_ADDR_REG] & 0xffffff;
+    unsigned int memaddr = dma->memaddr & 0xfff;
+    unsigned int dramaddr = dma->dramaddr & 0xffffff;
 
-    unsigned char *spmem = (unsigned char*)sp->mem + (sp->regs[SP_MEM_ADDR_REG] & 0x1000);
+    unsigned char *spmem = (unsigned char*)sp->mem + (dma->memaddr & 0x1000);
     unsigned char *dram = (unsigned char*)sp->ri->rdram->dram;
 
```

Comment:
Do a single if(dma->dir == SP_DMA_READ) at the top of the outermost for loop. Otherwise you get the branch penalty at each iteration. Something like:

```
if (dma->dir == SP_DMA_READ) {
// full dma read logic (no extra if)
}
else {
// full dma write logic (no extra if)
}
```

Note: If you remove the SP_NO_DMA, SP_DMA_READ will equal 0 and therefore might be better to compare against than SP_DMA_WRITE.